File: client_core/anymmkv/ios/Classes/AnymmkvPlugin.m
#import "AnymmkvPlugin.h"
#import <MMKV/MMKV.h>


#define ID @"id"
#define MMAPID @"mmapid"
#define CRYPT @"crypt"
#define KEY @"key"
#define VALUE @"value"

static inline BOOL isEmpty(id thing) {
  return thing == nil || [thing isKindOfClass:[NSNull class]] ||
         ([thing isKindOfClass:[NSData class]] &&
          [(NSData *)thing length] == 0) ||
         ([thing isKindOfClass:[NSArray class]] &&
          [(NSArray *)thing count] == 0) ||
         ([thing isKindOfClass:[NSString class]] &&
          [(NSString *)thing length] == 0);
}

@implementation AnymmkvPlugin {
  NSMutableArray *_mmkvs;
}

- (id)init {
  _mmkvs = [[NSMutableArray alloc] initWithCapacity:4];
  return self;
}

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar> *)registrar {
  FlutterMethodChannel *channel =
      [FlutterMethodChannel methodChannelWithName:@"anychat.github.com/anymmkv"
                                  binaryMessenger:[registrar messenger]];
  AnymmkvPlugin *instance = [[AnymmkvPlugin alloc] init];
  [registrar addMethodCallDelegate:instance channel:channel];
}

- (void)handleMethodCall:(FlutterMethodCall *)call
                  result:(FlutterResult)result {
  NSString *method = [call method];
  NSDictionary *arguments = [call arguments];

  MMKV *mmkv;
  if ([method isEqualToString:@"init"]) {
    NSString *mmapID = arguments[MMAPID];
    NSString *crypt = arguments[CRYPT];

    NSLog(@"arguments:%@", arguments);
    //  @();
    int rtv = (int)_mmkvs.count;

    if (isEmpty(mmapID)) {
      NSLog(@"xxxx1\n");
      mmkv = [MMKV defaultMMKV];
    } else {
      if (isEmpty(crypt)) {
      NSLog(@"xxxx2\n");
        mmkv = [MMKV mmkvWithID:mmapID];
      } else {
      NSLog(@"xxxx3\n");
        mmkv = [MMKV mmkvWithID:mmapID
                       cryptKey:[crypt dataUsingEncoding:NSUTF8StringEncoding]];
      }
    }
    [_mmkvs addObject:mmkv];
    result([NSNumber numberWithLong:rtv]);
    return;
  }
///////////////////////////////////////////
  int idx = ((NSNumber *)arguments[ID]).intValue;
  if (idx < 0 || idx > _mmkvs.count) {
    result([FlutterError errorWithCode:@"MMKVException"
                               message:@"MMKV index is out of range."
                               details:nil]);
    return;
  }


/////////////////////////////////////////////
  mmkv = _mmkvs[idx];
  NSString *key = arguments[KEY];
  NSObject *value = arguments[VALUE];

  if (isEmpty(mmkv)) {
    result([FlutterError errorWithCode:@"MMKVException"
                               message:@"MMKV is null."
                               details:nil]);
    return;
  }



  /////////////////////////////////////
  if ([method isEqualToString:@"setBool"]) {
    result([NSNumber numberWithBool:[mmkv setBool:((NSNumber *)value).boolValue
                                               forKey:key]]);
  } else if ([method isEqualToString:@"getBool"]) {
    if ([mmkv containsKey:key]) {
      result([NSNumber numberWithBool:[mmkv getBoolForKey:key]]);
    } else {
      result(nil);
    }
  } else if ([method isEqualToString:@"setInt"]) {
    result([NSNumber numberWithBool:[mmkv setInt64:((NSNumber *)value).longValue
                                            forKey:key]]);
  } else if ([method isEqualToString:@"getInt"]) {
    if ([mmkv containsKey:key]) {
      result([NSNumber numberWithLong:[mmkv getInt64ForKey:key]]);
    } else {
      result(nil);
    }
  } else if ([method isEqualToString:@"setDouble"]) {
    result(
        [NSNumber numberWithBool:[mmkv setDouble:((NSNumber *)value).doubleValue
                                          forKey:key]]);
  } else if ([method isEqualToString:@"getDouble"]) {
    if ([mmkv containsKey:key]) {
      result([NSNumber numberWithDouble:[mmkv getDoubleForKey:key]]);
    } else {
      result(nil);
    }
  } else if ([method isEqualToString:@"setString"]) {
    result([NSNumber numberWithBool:[mmkv setString:(NSString *)value
                                             forKey:key]]);
  } else if ([method isEqualToString:@"getString"]) {
    result([mmkv getStringForKey:key]);
  } else if ([method isEqualToString:@"setBytes"]) {
    result([NSNumber
        numberWithBool:[mmkv setData:((FlutterStandardTypedData *)value).data
                              forKey:key]]);
  } else if ([method isEqualToString:@"getBytes"]) {
    result([mmkv getDataForKey:key]);
  } else if ([method isEqualToString:@"contains"]) {
    result([NSNumber numberWithBool:[mmkv containsKey:key]]);
  } else if ([method isEqualToString:@"removeByKey"]) {
    [mmkv removeValueForKey:key];
    result(@YES);
  } else if ([method isEqualToString:@"clear"]) {
    [mmkv clearAll];
    result(@YES);
  } else if ([method isEqualToString:@"count"]) {
    result([NSNumber numberWithLong:[mmkv count]]);
  } else if ([method isEqualToString:@"allKeys"]) {
    result([mmkv allKeys]);
  } else if ([method isEqualToString:@"totalSize"]) {
    result([NSNumber numberWithLong:[mmkv totalSize]]);
  } else if ([method isEqualToString:@"getValueSize"]) {
    // getValueSizeForKey
    result([NSNumber numberWithLong:[mmkv getValueSizeForKey:key]]);
  } else {
    result(FlutterMethodNotImplemented);
  }
}

@end

File: client_core/anymmkv/macos/Classes/AnymmkvPlugin.swift
import Cocoa
import FlutterMacOS

import Foundation

import MMKV.MMKV

let ID = "id"
let MMAPID  = "mmapid"
let CRYPT = "crypt"
let KEY = "key"
let VALUE = "value"

public class AnymmkvPlugin: NSObject, FlutterPlugin {
    let mmkvrootdir = getApplicationSupportDirectory() ;
    var _mmkvs = [MMKV]()
    
    public static func register(with registrar: FlutterPluginRegistrar) {

        let channel = FlutterMethodChannel(name: "anychat.github.com/anymmkv", binaryMessenger: registrar.messenger)
        let instance = AnymmkvPlugin()
        registrar.addMethodCallDelegate(instance, channel: channel)
    }
  
    public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "init":
            let arguments = call.arguments as! [String: Any]
            let mmapid: String? = arguments[MMAPID] as? String
            let crypt: String? = arguments[CRYPT] as? String
            
            let rtv = _mmkvs.count
                if(mmkvrootdir != nil){
                    let mkvpath = "/mmkv";
                    let rootpatch = mmkvrootdir! + mkvpath;
                    
                   MMKV.initialize(rootDir:rootpatch);
                }
            if mmapid != nil{
                if crypt != nil{
                    let mmkv = MMKV.init(mmapID: mmapid!, cryptKey: crypt!.data(using: String.Encoding.utf8));
                    _mmkvs.append(mmkv!)
                }else{
                    let mmkv = MMKV.init(mmapID: mmapid!)
                    _mmkvs.append(mmkv!)
                }
            }else{
                let mmkv = MMKV.init()
                _mmkvs.append(mmkv)
            }
            
            result(rtv)
            return
        case "setBool":
            let arguments = call.arguments as! [String: Any]
            let idx = arguments[ID] as! Int
            if idx<0||idx>_mmkvs.count {
                result("MMKVException"+"MMKV index is out of range.")
                return
            }
            let mmkv = _mmkvs[idx]
            if mmkv == nil {
                result("MMKVException"+"MMKV is null")
                return
            }
            
            let key = arguments[KEY] as! String
            let value = arguments[VALUE] as! Bool
            
            result(mmkv.set(value, forKey: key))
        case "getBool":
            let arguments = call.arguments as! [String: Any]
            let idx = arguments[ID] as! Int
            if idx<0||idx>_mmkvs.count {
                result("MMKVException"+"MMKV index is out of range.")
                return
            }
            let mmkv = _mmkvs[idx]
            if mmkv == nil {
                result("MMKVException"+"MMKV is null")
                return
            }
            
            let key = arguments[KEY] as! String
            
            result(mmkv.bool(forKey: key))
        case "setInt":
            let arguments = call.arguments as! [String: Any]
            let idx = arguments[ID] as! Int
            if idx<0||idx>_mmkvs.count {
                result("MMKVException"+"MMKV index is out of range.")
                return
            }
            let mmkv = _mmkvs[idx]
            if mmkv == nil {
                result("MMKVException"+"MMKV is null")
                return
            }
            
            let key = arguments[KEY] as! String
            let value = arguments[VALUE] as! Int64
            
            result(mmkv.set(value, forKey: key))
        case "getInt":
            let arguments = call.arguments as! [String: Any]
            let idx = arguments[ID] as! Int
            if idx<0||idx>_mmkvs.count {
                result("MMKVException"+"MMKV index is out of range.")
                return
            }
            let mmkv = _mmkvs[idx]
            if mmkv == nil {
                result("MMKVException"+"MMKV is null")
                return
            }
            
            let key = arguments[KEY] as! String
            
            result(mmkv.int64(forKey: key))
        case "setDouble":
            let arguments = call.arguments as! [String: Any]
            let idx = arguments[ID] as! Int
            if idx<0||idx>_mmkvs.count {
                result("MMKVException"+"MMKV index is out of range.")
                return
            }
            let mmkv = _mmkvs[idx]
            if mmkv == nil {
                result("MMKVException"+"MMKV is null")
                return
            }
            
            let key = arguments[KEY] as! String
            let value = arguments[VALUE] as! Double
            
            result(mmkv.set(value, forKey: key))
        case "getDouble":
            let arguments = call.arguments as! [String: Any]
            let idx = arguments[ID] as! Int
            if idx<0||idx>_mmkvs.count {
                result("MMKVException"+"MMKV index is out of range.")
                return
            }
            let mmkv = _mmkvs[idx]
            if mmkv == nil {
                result("MMKVException"+"MMKV is null")
                return
            }
            
            let key = arguments[KEY] as! String
            
            result(mmkv.double(forKey: key))
        case "setString":
            let arguments = call.arguments as! [String: Any]
            let idx = arguments[ID] as! Int
            if idx<0||idx>_mmkvs.count {
                result("MMKVException"+"MMKV index is out of range.")
                return
            }
            let mmkv = _mmkvs[idx]
            if mmkv == nil {
                result("MMKVException"+"MMKV is null")
                return
            }
            
            let key = arguments[KEY] as! String
            let value = arguments[VALUE] as! String
            
            result(mmkv.set(value, forKey: key))
        case "getString":
            let arguments = call.arguments as! [String: Any]
            let idx = arguments[ID] as! Int
            if idx<0||idx>_mmkvs.count {
                result("MMKVException"+"MMKV index is out of range.")
                return
            }
            let mmkv = _mmkvs[idx]
            if mmkv == nil {
                result("MMKVException"+"MMKV is null")
                return
            }
            
            let key = arguments[KEY] as! String
            
            result(mmkv.string(forKey: key))
        case "setBytes":
            let arguments = call.arguments as! [String: Any]
            let idx = arguments[ID] as! Int
            if idx<0||idx>_mmkvs.count {
                result("MMKVException"+"MMKV index is out of range.")
                return
            }
            let mmkv = _mmkvs[idx]
            if mmkv == nil {
                result("MMKVException"+"MMKV is null")
                return
            }
            
            let key = arguments[KEY] as! String
            let value = arguments[VALUE] as! FlutterStandardTypedData
            
            result(mmkv.set(value.data, forKey: key))
        case "getBytes":
            let arguments = call.arguments as! [String: Any]
            let idx = arguments[ID] as! Int
            if idx<0||idx>_mmkvs.count {
                result("MMKVException"+"MMKV index is out of range.")
                return
            }
            let mmkv = _mmkvs[idx]
            if mmkv == nil {
                result("MMKVException"+"MMKV is null")
                return
            }
            
            let key = arguments[KEY] as! String
            
            result(mmkv.data(forKey: key))
        case "contains":
            let arguments = call.arguments as! [String: Any]
            let idx = arguments[ID] as! Int
            if idx<0||idx>_mmkvs.count {
                result("MMKVException"+"MMKV index is out of range.")
                return
            }
            let mmkv = _mmkvs[idx]
            if mmkv == nil {
                result("MMKVException"+"MMKV is null")
                return
            }
            
            let key = arguments[KEY] as! String
            
            result(mmkv.contains(key: key))
            
        case "removeByKey":
            let arguments = call.arguments as! [String: Any]
            let idx = arguments[ID] as! Int
            if idx<0||idx>_mmkvs.count {
                result("MMKVException"+"MMKV index is out of range.")
                return
            }
            let mmkv = _mmkvs[idx]
            if mmkv == nil {
                result("MMKVException"+"MMKV is null")
                return
            }
            
            let key = arguments[KEY] as! String
            mmkv.removeValue(forKey: key)
            result(true)
        case "clear":
            let arguments = call.arguments as! [String: Any]
            let idx = arguments[ID] as! Int
            if idx<0||idx>_mmkvs.count {
                result("MMKVException"+"MMKV index is out of range.")
                return
            }
            let mmkv = _mmkvs[idx]
            if mmkv == nil {
                result("MMKVException"+"MMKV is null")
                return
            }
            mmkv.clearAll()
            result(true)
        case "count":
            let arguments = call.arguments as! [String: Any]
            let idx = arguments[ID] as! Int
            if idx<0||idx>_mmkvs.count {
                result("MMKVException"+"MMKV index is out of range.")
                return
            }
            let mmkv = _mmkvs[idx]
            if mmkv == nil {
                result("MMKVException"+"MMKV is null")
                return
            }
            
            result(mmkv.count())
        case "allKeys":
            let arguments = call.arguments as! [String: Any]
            let idx = arguments[ID] as! Int
            if idx<0||idx>_mmkvs.count {
                result("MMKVException"+"MMKV index is out of range.")
                return
            }
            let mmkv = _mmkvs[idx]
            if mmkv == nil {
                result("MMKVException"+"MMKV is null")
                return
            }
            
            result(mmkv.allKeys())
        case "totalSize":
            let arguments = call.arguments as! [String: Any]
            let idx = arguments[ID] as! Int
            if idx<0||idx>_mmkvs.count {
                result("MMKVException"+"MMKV index is out of range.")
                return
            }
            let mmkv = _mmkvs[idx]
            if mmkv == nil {
                result("MMKVException"+"MMKV is null")
                return
            }
            
            result(mmkv.totalSize())
        case "getValueSize":
            let arguments = call.arguments as! [String: Any]
            let idx = arguments[ID] as! Int
            if idx<0||idx>_mmkvs.count {
                result("MMKVException"+"MMKV index is out of range.")
                return
            }
            let mmkv = _mmkvs[idx]
            if mmkv == nil {
                result("MMKVException"+"MMKV is null")
                return
            }
            
            let key = arguments[KEY] as! String
            
            result(mmkv.valueSize(forKey: key))
        default:
            result(FlutterMethodNotImplemented)
        }
    }  
}
private func getApplicationSupportDirectory() -> String? {
      var path = getDirectory(ofType: FileManager.SearchPathDirectory.applicationSupportDirectory)
      if let basePath = path {
        let basePathURL = URL.init(fileURLWithPath: basePath)
        path = basePathURL.appendingPathComponent(Bundle.main.bundleIdentifier!).path
        do {
          try FileManager.default.createDirectory(atPath: path!, withIntermediateDirectories: true)
        } catch {
          return nil
        }
      }
      return path
}

private func getDirectory(ofType directory: FileManager.SearchPathDirectory) -> String? {
  let paths = NSSearchPathForDirectoriesInDomains(
    directory,
    FileManager.SearchPathDomainMask.userDomainMask,
    true)
  return paths.first
}
